src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,335 @@
+/-
+Copyright (c) 2021 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Eric Wieser
+-/
+
+import ring_theory.ideal.basic
+import ring_theory.ideal.operations
+import linear_algebra.finsupp
+import ring_theory.graded_algebra.basic
+
+/-!
+
+# Homogeneous ideal of a graded algebra
+
+This file defines homogeneous ideals of `graded_algebra A` where `A : ι → submodule R R`and
+operations on them:
+* `mul`, `inf`, `Inf` of homogeneous ideals are homogeneous;
+* `⊤`, `⊥`, i.e. the trivial ring and `R` are homogeneous;
+* `radical` of a homogeneous ideal is homogeneous.
+-/
+
+noncomputable theory
+
+section is_homogeneous_ideal_defs
+
+open set_like direct_sum set
+open_locale big_operators
+
+variables {ι R : Type*} [comm_ring R] [decidable_eq ι] [add_comm_monoid ι]
+variables (A : ι → submodule R R) [graded_algebra A]
+variable (I : ideal R)
+
+/--An `I : ideal R` is called homogeneous if for every `r ∈ I`, every homogeneous component of `r`
+  is in `I`.-/
+@[reducible] def is_homogeneous_ideal : Prop :=
+  ∀ {i : ι} {r : R}, r ∈ I → (graded_algebra.decompose A r i : R) ∈ I
+
+lemma is_homogeneous_ideal_iff_forall_subset (I : ideal R) :
+  is_homogeneous_ideal A I ↔ ∀ i, (I : set R) ⊆ graded_algebra.proj A i ⁻¹' I :=
+iff.rfl
+
+lemma is_homogeneous_ideal_iff_subset_Inter (I : ideal R) :
+  is_homogeneous_ideal A I ↔ (I : set R) ⊆ ⋂ i, graded_algebra.proj A i ⁻¹' ↑I :=
+subset_Inter_iff.symm
+
+lemma is_homogeneous_ideal.exists_iff_eq_span :
+  (∃ (S : set (homogeneous_submonoid A)), I = ideal.span (coe '' S)) ↔
+  I = ideal.span {x | x ∈ I ∧ is_homogeneous A x} :=
+⟨λ ⟨S, hI⟩, begin
+    ext r, split; intro hr,
+  { rw hI at hr,
+    suffices : coe '' S ⊆ {x | x ∈ I ∧ is_homogeneous A x},
+    exact (ideal.span_mono this) hr,
+    intros s hs, split, rw hI,
+    refine ideal.subset_span hs,
+    obtain ⟨⟨s', homs'⟩, hs₁, hs₂⟩ := hs,
+    convert homs', rw ←hs₂, refl },
+  { obtain ⟨l, hl⟩ := (finsupp.mem_span_iff_total R _ _).mp hr,
+    rw ←hl, apply ideal.sum_mem, rintros ⟨x, hx₁, hx₂⟩ hx₃,
+    simp only [linear_map.id_coe, id.def, finsupp.mem_support_iff, linear_map.coe_smul_right,
+      ne.def, smul_eq_mul, subtype.coe_mk] at hx₁ hx₂ hx₃ ⊢,
+    exact ideal.mul_mem_left _ _ hx₁, }
+  end, λ hI, ⟨{x : homogeneous_submonoid A | ↑x ∈ I}, begin
+    rw hI, congr, ext r, split; intros hr,
+    { rcases hr with ⟨r_mem, ⟨i, r_eq⟩⟩,
+      use r, exact ⟨i, r_eq⟩, refine ⟨_, rfl⟩,
+      simp only [mem_set_of_eq, subtype.coe_mk], convert ←r_mem, },
+    { rcases hr with ⟨⟨r', hr'⟩, hr₁, hr₂⟩,
+      simp only [mem_set_of_eq, subtype.coe_mk] at hr₁,
+      rw ←hr₂, rw ←hI at hr₁, refine ⟨hr₁, hr'⟩, }
+  end⟩⟩
+
+variable [Π (i : ι) (x : A i), decidable (x ≠ 0)]
+
+lemma is_homogeneous_ideal.mul_homogeneous_element
+  {I : ideal R} (r x : R) (hx₁ : is_homogeneous A x) (hx₂ : x ∈ I) (j : ι) :
+  graded_algebra.proj A j (r * x) ∈ I :=
+begin
```

Comment:
This isn't about `is_homogeneous_ideal`, so doesn't really belong in the `is_homogeneous_ideal` namespace.